File: utils/restaurant.py
from utils.firebase import write_data, read_data, update_data, delete_data
from utils.category import add_restaurant_to_category, remove_restaurant_from_category

# moved
def create_restaurant(category_id, restaurant_id, data):
    """
    Add a restaurant and associate it with a category.
    :param category_id: The category ID to associate the restaurant with.
    :param restaurant_id: The unique ID for the restaurant.
    :param data: Dictionary containing restaurant details.
    """
    required_fields = [
        "name",
        "rating_1", 
        "rating_2",
        "rating_3",
        "notes",
        "visits", # list of date strings
        "location",
        "dishes",
        "photo"
    ]

    for field in required_fields:
        if field not in data:
            raise ValueError(f"Missing required field in restaurant creation: {field}")

    # Validate ratings
    for rating_field in ["rating_1", "rating_2", "rating_3"]:
        if not isinstance(data[rating_field], (int, float)) or not (0 <= data[rating_field] <= 10):
            raise ValueError(f"{rating_field} must be a number between 0 and 10.")

    # Add category reference to restaurant data
    restaurant_data = data.copy()
    restaurant_data["category_code"] = category_id

    path = f"restaurants/{restaurant_id}"
    write_data(path, restaurant_data)

    # Add restaurant to category, on the category-side
    add_restaurant_to_category(category_id, restaurant_id)

    return restaurant_data

# moved
def edit_rating(restaurant_id, rating_number, rating_value):
    """
    Edit a specific rating of a restaurant.
    :param restaurant_id: The restaurant ID to update.
    :param rating_number: Integer 1-3 specifying which rating to update.
    :param rating_value: The new rating value.
    """
    if not isinstance(rating_number, int) or rating_number not in [1, 2, 3]:
        raise ValueError("rating_number must be 1, 2, or 3")

    if not isinstance(rating_value, (int, float)) or not (0 <= rating_value <= 10):
        raise ValueError("rating_value must be a number between 0 and 10")

    rating_field = f"rating_{rating_number}"
    data = {rating_field: rating_value}

    path = f"restaurants/{restaurant_id}"
    update_data(path, data)
    return data

# moved
def edit_notes(restaurant_id, notes):
    """
    Edit the notes for a restaurant.
    :param restaurant_id: The restaurant ID to update.
    :param notes: String containing the new notes.
    """
    if not isinstance(notes, str):
        raise ValueError("notes must be a string")

    data = {"notes": notes}
    path = f"restaurants/{restaurant_id}"
    update_data(path, data)
    return data

# moved
def add_visit(restaurant_id, visit_date):
    """
    Add a visit date to a restaurant.
    :param restaurant_id: The restaurant ID to update.
    :param visit_date: String containing the visit date.
    """
    if not isinstance(visit_date, str):
        raise ValueError("visit_date must be a string")

    path = f"restaurants/{restaurant_id}/visits"
    visits = read_data(path) or []
    visits.append(visit_date)
    write_data(path, visits)
    return visits

# moved
def delete_visit(restaurant_id):
    """
    Delete the most recent visit date from a restaurant.
    :param restaurant_id: The restaurant ID to update.
    """
    path = f"restaurants/{restaurant_id}/visits"
    visits = read_data(path) or []
    if visits:
        visits.pop()
        write_data(path, visits)
    return visits

# moved
def edit_dish_ranking(restaurant_id, dishes):
    """
    Edit the dishes for a restaurant.
    :param restaurant_id: The restaurant ID to update.
    :param dishes: Array containing the new ordered list of dishes.
    """
    if not isinstance(dishes, list):
        raise ValueError("dishes must be an array")

    path = f"restaurants/{restaurant_id}/dishes"
    write_data(path, dishes)
    return dishes

# moved
def add_dish(restaurant_id, dish):
    """
    Add a dish to a restaurant.
    :param restaurant_id: The restaurant ID to update.
    :param dish: String containing the dish name to add.
    """
    if not isinstance(dish, str):
        raise ValueError("dish must be a string")

    path = f"restaurants/{restaurant_id}/dishes"
    dishes = read_data(path) or []
    dishes.append(dish)
    write_data(path, dishes)
    return dishes

# moved
def delete_dish(restaurant_id, dish):
    """
    Delete a dish from a restaurant.
    :param restaurant_id: The restaurant ID to update.
    :param dish: String containing the dish name to delete.
    """
    if not isinstance(dish, str):
        raise ValueError("dish must be a string")

    path = f"restaurants/{restaurant_id}/dishes"
    dishes = read_data(path) or []
    dishes.remove(dish)
    write_data(path, dishes)
    return dishes

def delete_restaurant(restaurant_id):
    """
    Delete a restaurant.
    :param restaurant_id: The restaurant ID to delete.
    """
    path = f"restaurants/{restaurant_id}"

    # Remove restaurant from category, on the category-side
    restaurant_data = read_data(path)
    if restaurant_data and "category_code" in restaurant_data:
        remove_restaurant_from_category(restaurant_data["category_code"], restaurant_id)

    delete_data(path)

# moved
def switch_restaurant_category(restaurant_id, new_category_id):
    """
    Change a restaurant's category.
    :param restaurant_id: The restaurant ID to update.
    :param new_category_id: String containing the new category ID.
    """
    if not isinstance(new_category_id, str):
        raise ValueError("Category ID must be a string")

    path = f"restaurants/{restaurant_id}"
    restaurant_data = read_data(path)
    
    if not restaurant_data:
        raise ValueError(f"Restaurant {restaurant_id} not found")
        
    restaurant_data["category_code"] = new_category_id
    write_data(path, restaurant_data)

    # Remove restaurant from old category, on the category-side
    remove_restaurant_from_category(restaurant_id, restaurant_data["category_code"])

    # Add restaurant to new category, on the category-side
    add_restaurant_to_category(new_category_id, restaurant_id)

    return restaurant_data

File: functions/restaurant.js
const admin = require("firebase-admin");
const {addRestaurantToCategory, removeRestaurantFromCategory} = require("./category");

/**
 * Add a restaurant and associate it with a category.
 * @param {string} categoryId - The category ID to associate the restaurant with.
 * @param {string} restaurantId - The unique ID for the restaurant.
 * @param {Object} data - Object containing restaurant details.
 * @param {string} data.name - The name of the restaurant.
 * @param {number} data.rating_1 - First rating (0-10).
 * @param {number} data.rating_2 - Second rating (0-10).
 * @param {number} data.rating_3 - Third rating (0-10).
 * @param {string} data.notes - Notes about the restaurant.
 * @param {Array<string>} data.visits - List of visit dates.
 * @param {string} data.location - Restaurant location.
 * @param {Array<string>} data.dishes - List of dishes.
 * @param {string} data.photo - URL to restaurant photo.
 * @return {Promise<Object>} The created restaurant data.
 * @throws {Error} If required fields are missing or invalid.
 */
async function createRestaurant(categoryId, restaurantId, data) {
  try {
    const requiredFields = [
      "name",
      "rating_1",
      "rating_2",
      "rating_3",
      "notes",
      "visits",
      "location",
      "dishes",
      "photo",
    ];

    // Check for required fields
    for (const field of requiredFields) {
      if (!(field in data)) {
        throw new Error(`Missing required field in restaurant creation: ${field}`);
      }
    }

    // Validate ratings
    const ratingFields = ["rating_1", "rating_2", "rating_3"];
    for (const field of ratingFields) {
      const rating = data[field];
      if (typeof rating !== "number" || rating < 0 || rating > 10) {
        throw new Error(`${field} must be a number between 0 and 10`);
      }
    }

    // Add category reference to restaurant data
    const restaurantData = {
      ...data,
      category_code: categoryId,
    };

    // Write restaurant data
    const path = `restaurants/${restaurantId}`;
    await admin.database().ref(path).set(restaurantData);
    console.log(`Restaurant ${restaurantId} created successfully`);

    // Add restaurant to category
    await addRestaurantToCategory(categoryId, restaurantId);

    return restaurantData;
  } catch (error) {
    console.error(`Error creating restaurant: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Edit a specific rating of a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {number} ratingNumber - Integer 1-3 specifying which rating to update.
 * @param {number} ratingValue - The new rating value.
 * @return {Promise<Object>} Object containing the updated rating field and value.
 * @throws {Error} If rating number or value is invalid.
 */
async function editRating(restaurantId, ratingNumber, ratingValue) {
  try {
    // Validate rating number
    if (!Number.isInteger(ratingNumber) || ![1, 2, 3].includes(ratingNumber)) {
      throw new Error("rating_number must be 1, 2, or 3");
    }

    // Validate rating value
    if (typeof ratingValue !== "number" || ratingValue < 0 || ratingValue > 10) {
      throw new Error("rating_value must be a number between 0 and 10");
    }

    const ratingField = `rating_${ratingNumber}`;
    const data = {[ratingField]: ratingValue};

    // Update the rating
    const path = `restaurants/${restaurantId}`;
    await admin.database().ref(path).update(data);
    console.log(`Rating ${ratingNumber} updated for restaurant ${restaurantId}`);

    return data;
  } catch (error) {
    console.error(`Error updating rating: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Edit the notes for a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {string} notes - String containing the new notes.
 * @return {Promise<Object>} Object containing the updated notes.
 * @throws {Error} If notes is not a string.
 */
async function editNotes(restaurantId, notes) {
  try {
    if (typeof notes !== "string") {
      throw new Error("notes must be a string");
    }

    const data = {notes};
    const path = `restaurants/${restaurantId}`;
    await admin.database().ref(path).update(data);
    console.log(`Notes updated for restaurant ${restaurantId}`);

    return data;
  } catch (error) {
    console.error(`Error updating notes: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Add a visit date to a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {string} visitDate - String containing the visit date.
 * @return {Promise<Array<string>>} Updated list of visit dates.
 * @throws {Error} If visitDate is not a string.
 */
async function addVisit(restaurantId, visitDate) {
  try {
    if (typeof visitDate !== "string") {
      throw new Error("visit_date must be a string");
    }

    const path = `restaurants/${restaurantId}/visits`;
    const snapshot = await admin.database().ref(path).get();
    const visits = snapshot.val() || [];

    visits.push(visitDate);
    await admin.database().ref(path).set(visits);
    console.log(`Visit date added to restaurant ${restaurantId}`);

    return visits;
  } catch (error) {
    console.error(`Error adding visit date: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Delete the most recent visit date from a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @return {Promise<Array<string>>} Updated list of visit dates.
 */
async function deleteVisit(restaurantId) {
  try {
    const path = `restaurants/${restaurantId}/visits`;
    const snapshot = await admin.database().ref(path).get();
    const visits = snapshot.val() || [];

    if (visits.length > 0) {
      visits.pop();
      await admin.database().ref(path).set(visits);
      console.log(`Most recent visit deleted from restaurant ${restaurantId}`);
    }

    return visits;
  } catch (error) {
    console.error(`Error deleting visit date: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Edit the dishes for a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {Array<string>} dishes - Array containing the new ordered list of dishes.
 * @return {Promise<Array<string>>} Updated list of dishes.
 * @throws {Error} If dishes is not an array.
 */
async function editDishRanking(restaurantId, dishes) {
  try {
    if (!Array.isArray(dishes)) {
      throw new Error("dishes must be an array");
    }

    const path = `restaurants/${restaurantId}/dishes`;
    await admin.database().ref(path).set(dishes);
    console.log(`Dishes updated for restaurant ${restaurantId}`);

    return dishes;
  } catch (error) {
    console.error(`Error updating dishes: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Add a dish to a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {string} dish - String containing the dish name to add.
 * @return {Promise<Array<string>>} Updated list of dishes.
 * @throws {Error} If dish is not a string.
 */
async function addDish(restaurantId, dish) {
  try {
    if (typeof dish !== "string") {
      throw new Error("dish must be a string");
    }

    const path = `restaurants/${restaurantId}/dishes`;
    const snapshot = await admin.database().ref(path).get();
    const dishes = snapshot.val() || [];

    dishes.push(dish);
    await admin.database().ref(path).set(dishes);
    console.log(`Dish added to restaurant ${restaurantId}`);

    return dishes;
  } catch (error) {
    console.error(`Error adding dish: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Delete a dish from a restaurant.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {string} dish - String containing the dish name to delete.
 * @return {Promise<Array<string>>} Updated list of dishes.
 * @throws {Error} If dish is not a string or not found.
 */
async function deleteDish(restaurantId, dish) {
  try {
    if (typeof dish !== "string") {
      throw new Error("dish must be a string");
    }

    const path = `restaurants/${restaurantId}/dishes`;
    const snapshot = await admin.database().ref(path).get();
    const dishes = snapshot.val() || [];

    const index = dishes.indexOf(dish);
    if (index === -1) {
      throw new Error(`Dish "${dish}" not found in restaurant ${restaurantId}`);
    }

    dishes.splice(index, 1);
    await admin.database().ref(path).set(dishes);
    console.log(`Dish deleted from restaurant ${restaurantId}`);

    return dishes;
  } catch (error) {
    console.error(`Error deleting dish: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Change a restaurant's category.
 * @param {string} restaurantId - The restaurant ID to update.
 * @param {string} newCategoryId - String containing the new category ID.
 * @return {Promise<Object>} Updated restaurant data.
 * @throws {Error} If category ID is not a string or restaurant not found.
 */
async function switchRestaurantCategory(restaurantId, newCategoryId) {
  try {
    if (typeof newCategoryId !== "string") {
      throw new Error("Category ID must be a string");
    }

    const path = `restaurants/${restaurantId}`;
    const snapshot = await admin.database().ref(path).get();
    const restaurantData = snapshot.val();

    if (!restaurantData) {
      throw new Error(`Restaurant ${restaurantId} not found`);
    }

    const oldCategoryId = restaurantData.category_code;
    restaurantData.category_code = newCategoryId;

    // Update restaurant data first
    await admin.database().ref(path).set(restaurantData);

    // Remove restaurant from old category
    await removeRestaurantFromCategory(oldCategoryId, restaurantId);

    // Add restaurant to new category
    await addRestaurantToCategory(newCategoryId, restaurantId);

    console.log(`Restaurant ${restaurantId} moved to category ${newCategoryId}`);
    return restaurantData;
  } catch (error) {
    console.error(`Error switching restaurant category: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

/**
 * Delete a restaurant.
 * @param {string} restaurantId - The restaurant ID to delete.
 * @return {Promise<void>}
 */
async function deleteRestaurant(restaurantId) {
  try {
    const path = `restaurants/${restaurantId}`;
    // Get restaurant data first
    const snapshot = await admin.database().ref(path).get();
    const restaurantData = snapshot.val();

    // Remove from category if it exists and has a category code
    if (restaurantData && restaurantData.category_code) {
      await removeRestaurantFromCategory(restaurantData.category_code, restaurantId);
    }

    // Delete the restaurant
    await admin.database().ref(path).remove();
    console.log(`Restaurant ${restaurantId} deleted successfully`);
  } catch (error) {
    console.error(`Error deleting restaurant: ${error.message}`);
    throw error; // Re-throw to maintain the error message
  }
}

module.exports = {
  createRestaurant,
  editRating,
  editNotes,
  addVisit,
  deleteVisit,
  editDishRanking,
  addDish,
  deleteDish,
  switchRestaurantCategory,
  deleteRestaurant,
};
